File: frontend/lib/stats.ts
import { apiFetch } from "./api";

export type MonthCount = { month: string; count: number };
export type GenreCount = { genre: string; count: number };
export type StarCount = { rating: number; count: number };
export type MemberStat = {
  user_id: number;
  nickname: string;
  picks: number;
  pages_read: number;
  avg_given: number | null;
};

export type GroupStats = {
  books_completed: number;
  pages_total: number;
  avg_rating: number | null;
  loops: number;
  by_month: MonthCount[];
  by_genre: GenreCount[];
  star_distribution: StarCount[];
  members: MemberStat[];
};

export function getStats(gid: number | string): Promise<GroupStats> {
  return apiFetch<GroupStats>(`/groups/${gid}/stats`);
}

File: backend/app/schemas/stats.py
from pydantic import BaseModel


class MonthCount(BaseModel):
    month: str
    count: int


class GenreCount(BaseModel):
    genre: str
    count: int


class StarCount(BaseModel):
    rating: int
    count: int


class MemberStat(BaseModel):
    user_id: int
    nickname: str
    picks: int
    pages_read: int
    avg_given: float | None


class GroupStats(BaseModel):
    books_completed: int
    pages_total: int
    avg_rating: float | None
    loops: int
    by_month: list[MonthCount]
    by_genre: list[GenreCount]
    star_distribution: list[StarCount]
    members: list[MemberStat]
